File: libs/power-log-parser/src/lib/handlers/change-entity-handler.ts
import type { Helper } from '../helper';
import { Action, ChangeEntity, Game, Node, NodeType, Tag } from '../models';
import { Regexes } from '../regexes';
import type { ParserState } from '../state/parser-state';

export class ChangeEntityHandler {
	static Handle(
		timestamp: string,
		data: string,
		state: ParserState,
		indentLevel: number,
		helper: Helper,
	): boolean {
		const match = Regexes.ActionChangeEntityRegex.exec(data);
		if (match) {
			const rawEntity = match[1];
			const cardId = match[2];
			const entity = helper.ParseEntity(rawEntity);

			const changeEntity = new ChangeEntity();
			changeEntity.CardId = cardId;
			changeEntity.Entity = entity;
			changeEntity.Tags = [];
			changeEntity.TimeStamp = timestamp;

			state.UpdateCurrentNode(NodeType.Game, NodeType.Action);
			if (state.Node!.Type === NodeType.Game) {
				(state.Node!.Object as Game).AddData(changeEntity);
			} else if (state.Node!.Type === NodeType.Action) {
				(state.Node!.Object as Action).Data.push(changeEntity);
			} else {
				throw new Error('Invalid node ' + state.Node!.Type);
			}
			const newNode = new Node(NodeType.ChangeEntity, changeEntity, indentLevel, state.Node, data);
			state.CreateNewNode(newNode);
			state.Node = newNode;
			return true;
		}
		return false;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Parser/Handlers/DataHandler/ChangeEntityHandler.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Parser.Handlers
{
    internal class ChangeEntityHandler
    {
        public static bool HandleChangeEntity(DateTime timestamp, string data, ParserState state, int indentLevel, Helper helper)
        {
            var match = Regexes.ActionChangeEntityRegex.Match(data);
            if (match.Success)
            {
                var rawEntity = match.Groups[1].Value;
                var cardId = match.Groups[2].Value;
                var entity = helper.ParseEntity(rawEntity);

                var changeEntity = new ChangeEntity { CardId = cardId, Entity = entity, Tags = new List<Tag>(), TimeStamp = timestamp };
                state.UpdateCurrentNode(typeof(Game), typeof(Action));
                if (state.Node.Type == typeof(Game))
                    ((Game)state.Node.Object).AddData(changeEntity);
                else if (state.Node.Type == typeof(Action))
                    ((Action)state.Node.Object).Data.Add(changeEntity);
                else
                    throw new Exception("Invalid node " + state.Node.Type);
                var newNode = new Node(typeof(ChangeEntity), changeEntity, indentLevel, state.Node, data);
                state.CreateNewNode(newNode);
                state.Node = newNode;
                return true;
            }
            return false;
        }
    }
}
